Lib/_pyrepl/reader.py

```diff
@@ -433,6 +543,190 @@ def eow(self, p: int | None = None) -> int:
+        of the next word or punctuation sequence."""
+        if p is None:
+            p = self.pos
+        b = self.buffer
+
+        if not b or p >= len(b):
+            return max(0, len(b) - 1) if b else 0
+
+        # Skip current word or punctuation sequence
+        if _is_vi_word_char(b[p]):
+            # On a word char - skip word chars
+            while p < len(b) and _is_vi_word_char(b[p]):
+                p += 1
+        elif not b[p].isspace():
+            # On punctuation - skip punctuation
+            while p < len(b) and not _is_vi_word_char(b[p]) and not b[p].isspace():
+                p += 1
+
+        # Skip whitespace to find next word or punctuation
+        while p < len(b) and b[p].isspace():
+            p += 1
+
+        # Clamp to valid buffer range
+        return min(p, len(b) - 1) if b else 0
+
+    def vi_forward_word_ws(self, p: int | None = None) -> int:
+        """Return the 0-based index of the first character of the next WORD
+        (vi 'W' semantics).
+
+        Treats white space as the only separator."""
+        if p is None:
+            p = self.pos
+        b = self.buffer
+
+        if not b or p >= len(b):
+            return max(0, len(b) - 1) if b else 0
+
+        # Skip all non-whitespace (the current WORD)
+        while p < len(b) and not b[p].isspace():
+            p += 1
+
+        # Skip whitespace to find next WORD
+        while p < len(b) and b[p].isspace():
+            p += 1
+
+        # Clamp to valid buffer range
+        return min(p, len(b) - 1) if b else 0
+
+    def vi_bow(self, p: int | None = None) -> int:
+        """Return the 0-based index of the beginning of the word preceding p
+        (vi 'b' semantics).
+
+        Vi has three character classes: word chars (alnum + _), punctuation
+        (non-word, non-whitespace), and whitespace. 'b' moves to the start
+        of the current or previous word/punctuation sequence."""
+        if p is None:
+            p = self.pos
+        b = self.buffer
+
+        if not b or p <= 0:
+            return 0
+
+        p -= 1
+
+        # Skip whitespace going backward
+        while p >= 0 and b[p].isspace():
+            p -= 1
+
+        if p < 0:
+            return 0
+
+        # Now skip the word or punctuation sequence we landed in
+        if _is_vi_word_char(b[p]):
+            while p > 0 and _is_vi_word_char(b[p - 1]):
+                p -= 1
+        else:
+            # Punctuation sequence
+            while p > 0 and not _is_vi_word_char(b[p - 1]) and not b[p - 1].isspace():
+                p -= 1
+
+        return p
+
+    def vi_bow_ws(self, p: int | None = None) -> int:
+        """Return the 0-based index of the beginning of the WORD preceding p
+        (vi 'B' semantics).
+
+        Treats white space as the only separator."""
+        if p is None:
+            p = self.pos
+        b = self.buffer
+
+        if not b or p <= 0:
+            return 0
+
+        p -= 1
+
```

Comment:
General comment: avoid blank lines and only use them to separate sections inside functions as per PEP-8. The stdlib is usually written compactly.